WIP

sujalshah-bit opened the pull request:
fixes: #14651.

<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

juliusv commented:
Before leaving the real review to @beorn7 the native histogram expert, a few quick observations / questions:

* Seems like the autocomplete for the new operator is still missing?
* Similar point for linting adjustments in the editor: an expression like `foo </ bool 1` currently gives no linter error but fails with an error like `bool modifier can only be used on comparison operators`. An expression like `1 </ 1` gives the linter error `comparisons between scalars must use BOOL modifier`, but the bool modifier does not help / is not legal in this case. I guess there needs to be just more differentiation between the regular comparison operators and the native histogram trimming ones.
* Should the `count` and `sum` fields stay completely untouched as they currently are? I guess you can't really change `sum` in a good way, and thus you also keep `count` without removing the filtered elements from it? It's just a bit funky to get a derived histogram that has potentially no buckets at all anymore, but a big count and sum. Not sure what to do about that though (unless we also stored a sum per bucket or estimated it by doing count * median).
* If I try running `non_histogram_metric </ 1`, I get the error `operator "</" not allowed for operations between Vectors`. Maybe the error message should be more fitting here, saying something about the operator only working on native histograms and scalars?

sujalshah-bit commented:
Thanks for the feedback, @juliusv! I've made the following changes:

- ✅ **Autocomplete** support for the new operator (`</`, `>/`) has been added.
  
- ✅ The **linter now properly distinguishes** native histogram trimming operators from standard comparison operators:
  - `foo </ bool 1` now triggers linting error:  
    `"bool modifier can only be used on comparison operators"`
  - `1 </ 1` now triggers linting error:  
    `"operator "</" not allowed for Scalar operations"`

- ✅ We’ve updated how the `sum` is handled for trimmed histograms:
  ```go
  updatedCount += bucket.Count
  bucketMidpoint := (bucket.Lower + bucket.Upper) / 2
  updatedSum += bucket.Count * bucketMidpoint
  ```
  This way, `sum` is now roughly estimated per bucket, and `count` reflects only the buckets that remain after trimming.

- ✅ The error message for invalid usage like `non_histogram_metric </ 1` now correctly states:  
  `"operator "</" is only supported for native histograms and scalars, not standard vector operations"`

Let me know if you have any further suggestions!

sujalshah-bit commented:
@juliusv is it ready for review for @beorn7 ?

beorn7 commented:
I'll have a look at this ASAP. Also, @NeerajGartia21 might help with the reviewing. He has worked a lot on histogram functions recently and has become one of the domain experts.

NeerajGartia21 commented on promql/engine.go:
```diff
@@ -2906,6 +2952,9 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
 		}
 	}
+	if parser.LTRIM == op || parser.RTRIM == op {
+		panic(fmt.Errorf("operator %q is only supported for native histograms and scalars, not standard vector operations", op))
```
We probably shouldn't panic in these cases. There could be situations where this operation is done between a histogram and a vector that has a mix of floats and histograms. In such cases, the user might expect partial results instead of a full failure. It would be better to return an info annotation here rather than panicking.

NeerajGartia21 commented on promql/engine.go:
```diff
@@ -2876,6 +2914,14 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
+			case parser.LTRIM:
+				trimmedHist := *hlhs
+				trimBuckets(&trimmedHist, hlhs, rhs, "LTRIM")
+				return 0, &trimmedHist, true, nil
```
We should also consider how this behaves when the `bool` modifier is used.  
@beorn7, can you please clarify what the expected output should be in this case?

beorn7 replied:
I don't think the `</` and `>/` operators should be allowed with the `bool` modifier. It's a syntax error.

NeerajGartia21 replied:
Yeah, so we should disable it. Currently, it does not throw any syntax error.

NeerajGartia21 commented on promql/engine.go:
```diff
@@ -2825,6 +2825,44 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+		}
+	}
+
+	for i, iter := 0, hlhs.NegativeBucketIterator(); iter.Next(); i++ {
```
Instead of iterating through positive and negative buckets separately, we could simplify the code by using `AllBucketIterator` to iterate over both at once. This might make the logic cleaner and easier to maintain.

NeerajGartia21 reviewed (commented):
The parser-side changes look good, but I think the trimming logic needs improvement:

1. 
   There's a mismatch between the operator name and its behavior. In https://github.com/prometheus/prometheus/issues/14651#issuecomment-2485736158 , @beorn7 mentioned that `histogram_metric </ 0.1` should remove observations greater than `0.1`. This means it trims the right part of the histogram, so logically it should be called `RTRIM`.  However, the PR description currently says that `</` removes observations below the threshold, which contradicts both the discussion and the current implementation. Would be good to clarify this, but I’ll leave the final word to @beorn7.

2. Right now, when the trim value lies between a bucket’s bounds, the entire bucket is discarded. But ideally, we should estimate (via interpolation) how many observations to remove from that bucket.  
   The current logic only works correctly when the bucket boundary exactly matches the trim parameter.  
   For example, if the trim value is `0.8`, schema is `0`, and all observations lie in `[0.5, 1)`, applying `LTRIM` currently gives no observations, but we should actually retain some.

   Also, make sure this works with native histograms with custom buckets, and handle zero buckets correctly as well.

For reference, you can look at the interpolation logic in the `histogram_quantile` and `histogram_fraction` functions [here](https://github.com/prometheus/prometheus/blob/main/promql/quantile.go#L203).

juliusv commented:
> 1. There's a mismatch between the operator name and its behavior. In [native histograms: Implement the </ and >/ operators to "trim" a histogram. #14651 (comment)](https://github.com/prometheus/prometheus/issues/14651#issuecomment-2485736158) , @beorn7 mentioned that `histogram_metric </ 0.1` should remove observations greater than `0.1`. This means it trims the right part of the histogram, so logically it should be called `RTRIM`.  However, the PR description currently says that `</` removes observations below the threshold, which contradicts both the discussion and the current implementation. Would be good to clarify this, but I’ll leave the final word to @beorn7.

While I can logically understand either variant, the second one feels way more intuitive to me when taking into account that traditional filter operators like `>` and `<` keep (and not drop) those series of a vector that match the filter condition. So I would expect `histogram_metric </ 0.1` to keep only the observations less than `0.1` and inverting it would be confusing.

NeerajGartia21 commented:
> While I can logically understand either variant, the second one feels way more intuitive to me when taking into account that traditional filter operators like > and < keep (and not drop) those series of a vector that match the filter condition. So I would expect histogram_metric </ 0.1 to keep only the observations less than 0.1 and inverting it would be confusing.

Makes sense. We could also call them `trimUpper` and `trimLower`.
- `</` as `trimUpper` (removes values > threshold)  
- `>/` as `trimLower` (removes values < threshold)  

Might make things clearer.

juliusv commented:
> > While I can logically understand either variant, the second one feels way more intuitive to me when taking into account that traditional filter operators like > and < keep (and not drop) those series of a vector that match the filter condition. So I would expect histogram_metric </ 0.1 to keep only the observations less than 0.1 and inverting it would be confusing.
> 
> Makes sense. We could also call them `trimUpper` and `trimLower`.
> 
> * `</` as `trimUpper` (removes values > threshold)
> * `>/` as `trimLower` (removes values < threshold)

Yeah, I think that kind of naming would make sense, it would mirror what e.g. Go does with `TrimLeft()` and `TrimRight()` (naming the part that is removed).

sujalshah-bit commented:
Just to confirm my understanding:

The intended behavior is that histogram_metric </ **rhs** keeps only values less than **rhs** (i.e., trims away values ≥ rhs), and >/ **rhs** keeps only values greater than **rhs** (i.e., trims away values ≤ rhs).
So the PR description needs to be corrected to match this behavior — is that right?

juliusv commented:
> Just to confirm my understanding:
> 
> The intended behavior is that histogram_metric </ **rhs** keeps only values less than **rhs** (i.e., trims away values ≥ rhs), and >/ **rhs** keeps only values greater than **rhs** (i.e., trims away values ≤ rhs). So the PR description needs to be corrected to match this behavior — is that right?

Correct!

beorn7 commented:
High level answers (I have not yet looked at the code):

WRT direction: I think the discussion above came to the right conclusion: `histogram </ threshold` means to filter out observation ≥  threshold and keep the observations < threshold (AKA `TrimUpper`), and vice versa for `>/`.

WRT interpolation: I believe we should do the interpolation. Linear interpolation for NHCB and in the zero bucket, and exponential interpolation for the standard exponential schemas. As @NeerajGartia21 said:

> For reference, you can look at the interpolation logic in the `histogram_quantile` and `histogram_fraction` functions [here](https://github.com/prometheus/prometheus/blob/main/promql/quantile.go#L203).

WRT `count` and `sum`: Those need to be adjusted as well, as @sujalshah-bit already discussed. To add more detail to the provided formula:
- For `count`, we need to deduct the bucket count of entirely removed buckets. For the bucket into which we have interpolated, we need to deduct the count we have removed from that bucket as per the interpolation.
- For `sum`, we can indeed use the midpoint of each bucket for linear interpolation (i.e. for NHCB and the zero bucket). For the exponential bucket, we should use the geometric mean (cf. [how it is done for `histogram_stddev`](https://github.com/prometheus/prometheus/blob/b4d3c06acb805f9b9e786eb268b7cea84b34d1b0/promql/functions.go#L1371)).

beorn7 commented:
@sujalshah-bit do you feel with the collected comments that you can go into the next iteration of working on the code?

My impression is that a code-level review only makes sense once you have worked in the high level notes above. If you need some code-level assistance, please let me know. (Note that @NeerajGartia21 already gave valuable code level comments.)

Also, of course, feel free to ask any questions you might still have to fully understand what we want to implement here.

beorn7 commented:
(Side note: When mentioning the geometric mean for `histogram_stddev` above, I realized that this function should take the arithmetic mean for NHCB. I filed https://github.com/prometheus/prometheus/issues/16439 for it.)

sujalshah-bit commented:
Thanks a lot for the detailed discussion above — I’ve learned a lot, and will be updating the changes soon!

I'm trying to **generate a customized bucket for native histograms** for testing this feature, but it's not working as expected.

By expected I mean:
```go
CustomValues []float64 // this is empty and,
FloatHistogram.UsesCustomBuckets() // this is returning false
```

Here's the code I'm using to define the histogram:

```go
customBuckets = promauto.NewHistogram(prometheus.HistogramOpts{
	Name: "custom_bucketss",
	Help: "Histogram with manually defined custom buckets",
	Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10},
	NativeHistogramBucketFactor: 1.1,
})
rng = rand.New(rand.NewPCG(uint64(time.Now().UnixNano()), 1))
customValue := rng.Float64() * 5.0
customBuckets.Observe(customValue)
```

Am I missing anything required to make it a proper **NHCB (Native Histogram with Custom Buckets)** for testing purposes?

beorn7 commented:
I would not create any histograms with client_golang.

Just use the PromQL testing framework. https://github.com/prometheus/prometheus/blob/main/promql/promqltest/testdata/native_histograms.test should give you loads if inspiration how to write test cases.

beorn7 commented:
https://github.com/prometheus/prometheus/blob/main/promql/promqltest/testdata/native_histograms.test also contains a bunch of NHCBs.

sujalshah-bit commented:
I’ve been feeling unwell due to the extensive heatwaves in my area lately, so I might need a bit more time to push the next set of changes.

Would it be okay if there’s a slight delay? I’ll try to get back to it as soon as I can.

NeerajGartia21 commented:
That’s totally fine. Your health comes first. Take the time you need, and I hope you feel better soon!

sujalshah-bit commented:
I’ve gone through and addressed the suggested changes across the board. Here's a summary of what has been done:

✅ **Interpolation for partial buckets**  
- The trimming logic now uses **linear interpolation** for the native histogram custom buckets and zero bucket when the threshold lies within a bucket, and **exponential interpolation** for exponential schemas, following the logic used in `histogram_quantile` and `histogram_fraction`.  
- This ensures partial buckets are no longer discarded outright and estimates are adjusted more accurately.

✅ **`Count` and `Sum` adjustment**  
- Implemented proper adjustment of `Count` and `Sum`:  
  - Full buckets beyond the trim threshold are excluded.  
  - Partial buckets contribute proportionally based on the interpolated fraction.  
  - For sum, used **midpoint** for linear buckets and **geometric mean** for exponential buckets.  
  - In partially trimmed buckets, used mean between the threshold and the nearest boundary (upper or lower depending on trim mode).


✅ **Refactorings and code quality** 
- Renamed `LTRIM` to `TRIM_UPPER` and `RTRIM` to `TRIM_LOWER`  
- Renamed `trimBuckets` to `trimHistogram` for better clarity and alignment with operator naming.  
- Removed the second histogram argument and now operate directly on a copied histogram (`hlhs.Copy()`), simplifying the function signature and reducing unnecessary copying.

✅ **Compaction**  
- Applied `.Compact(0)` to the histogram after trimming to remove any empty buckets.

✅ **Avoid panics**  
- Replaced the `panic` for unsupported TRIM_UPPER/TRIM_LOWER on mixed vectors with an **info annotation**, allowing for graceful fallback and partial evaluation in heterogeneous vectors.

✅ **Test cases**

Also updated PR description.
**WILL DO CLEANUPS AFTER THE CORE LOGIC GETS APPROVED**
Happy to iterate further based on feedback :)

beorn7 commented:
Thank you very much. I'll have a look ASAP.

beorn7 commented:
Apologies for delays. Company events and vacations in the way. I still plan to get to this eventually.
@NeerajGartia21 @krajorama this might be fun for you, too, if you feel bored. :)

NeerajGartia21 reviewed (commented):
The core logic looks solid overall, but there are a few important improvements to consider:

1. Currently, the approximate sum is calculated from the remaining buckets, which causes some deviation from the original sum since the previous sum isn’t utilized. Instead, I suggest calculating the approximate sum of the removed buckets and subtracting it from the actual sum. (We should handle edge cases carefully, for example, when only positive buckets exist but the resulting sum becomes negative.)

2. The current approach iterates through all buckets, but we can optimize by stopping once we’ve removed all relevant buckets, breaking the loop early.

3. The code is quite repetitive and somewhat confusing. To simplify, we could use `AllBucketIterator` or `AllReverseBucketIterator` based on the operation instead of handling positive, negative, and zero buckets separately. Also, since the main focus is on the bucket where the `trim` or `rhs` value lies, creating a helper function to calculate the fraction or count to remove from that bucket would clean up the logic. Unifying the fraction/count calculation for both custom and exponential buckets into a single function would make the code clearer and easier to maintain.

sujalshah-bit commented:
I will update the changes soon.

sujalshah-bit commented:
after this new method(to calculate the new SUM from the original SUM by deducting removed bucket's SUM) @NeerajGartia21 suggested all test fail. Do I have to calculate SUM manually for each histogram and pass it in when we adding data in test environment?

previously I was not passing SUM when adding histogram because we werer creating new SUM scratch so there was no need for it.

sujalshah-bit commented on web/ui/module/lezer-promql/src/promql.grammar:
```diff
@@ -78,6 +78,8 @@ BinaryExpr {
   expr !eql Gte    binModifiers expr |
   expr !eql Gtr    binModifiers expr |
+  expr !eql TrimUpper  binModifiers expr |
+  expr !eql TrimLower  binModifiers expr |
```
sure I will do it!

NeerajGartia21 commented:
> Do I have to calculate SUM manually for each histogram and pass it in when we adding data in test environment?

No, you don’t need to manually calculate and pass the `sum` for each histogram in the test environment. The `sum` field is automatically updated whenever observations are added to the histogram. It’s important to note that the `sum` is not derived from the average of bucket values; rather, it’s a separate field that tracks the total of all observations. When we need the sum for a subset of buckets (e.g., during trimming), we approximate it using bucket midpoints, but the actual `sum` field already exists in the histogram and should reflect the true total.

sujalshah-bit commented:
we're using the PromQL test framework, we can't call ` .Observe()` or simulate observations programmatically like we would in Go code. That means the sum field isn't auto-populated during tests — it's only what we explicitly provide in the load command.

Now that the new logic uses the actual sum value (i.e., original_sum - trimmed_buckets_sum), the sum must be set manually in the test data to avoid test failures. Otherwise, it defaults to zero or an incorrect value, which causes deviation from expected results.

Sorry if this is a dumb question, just wanted to be sure I understood it correctly before proceeding.

cc: @NeerajGartia21

beorn7 commented:
Sorry for late reply.

Indeed, for the tests via the framework (i.e. the changes in [promql/promqltest/testdata/native_histograms.test](https://github.com/prometheus/prometheus/pull/16330/files#diff-14c7f39d9c89d1873bad33d30a2e671d8bbcb90011856b21b7562b91831b50b2) in this PR), there is no `Observe` call. The input data in the `load` command has to be fully specified.

beorn7 commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1319,3 +1319,60 @@ eval instant at 10m histogram_sub_3{idx="0"} - ignoring(idx) histogram_sub_3{idx
+
+# Test native histogram with trim operators ("</": TRIM_UPPER, ">/": TRIM_LOWER)
+load 1m
+  h_test {{schema:0 count:34 z_bucket:1 z_bucket_w:0.001 buckets:[2 4 8 16] n_buckets:[1 2]}} 
+  h_test_2 {{schema:2 count:27 z_bucket:1 z_bucket_w:0.001 buckets:[1 2 4 7 3] n_buckets:[1 5 3 1]}}
+  cbh {{schema:-53 sum:5 count:15 custom_values:[5 10 15 20] buckets:[1 6 4 3 1]}}
+  zero_bucket {{schema:0 z_bucket:5 z_bucket_w:0.01 buckets:[2 3] n_buckets:[1 2 3]}}
```
To elaborate on my comment:

These should all have `sum:...` fields with plausible values.

`cbd` already has a `sum:...` field, but its value is not plausible. If we assume that the observations are more or less in the middle of each bucket (and around 2.5 for the first bucket and not much above 20 for the +Inf bucket), the sum should be around 170.

krajorama commented on promql/engine.go:
```diff
@@ -2887,6 +2887,334 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+				keepCount = bucket.Count
+			} else {
+				// Trim point is within bucket or below, keep none
+				keepCount = 0
```
We definitely need at least a  test case for this.

I can imagine two ways of handling this:
- just do the bucket.Count*0.5 as for +Inf.
- do bucket.Count*0.5  if the bucket.Upper is <= 0.0 otherwise assume this histogram is for positive values only (which we [assume](https://github.com/prometheus/prometheus/blob/2a88f562d12f309944cd5d0c462e3f6f5694dd15/promql/quantile.go#L148) in histogram_quantile as well) and treat bucket.Lower as 0, which means if rhs < 0 then keepCount == 0 otherwise 0 <= rhs < bucket.Upper so we can do linear interpolation.

krajorama commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1373,3 +1373,60 @@ eval instant at 1m histogram_fraction(-Inf, +Inf, histogram_nan)
+load 1m
+  h_test {{schema:0 sum:123.75 count:34 z_bucket:1 z_bucket_w:0.001 buckets:[2 4 8 16] n_buckets:[1 2]}} 
+  h_test_2 {{schema:2 sum:12.8286080906 count:28 z_bucket:1 z_bucket_w:0.001 buckets:[1 2 4 7 3] n_buckets:[1 5 3 1]}}
+  cbh {{schema:-53 sum:172.5 count:15 custom_values:[5 10 15 20] buckets:[1 6 4 3 1]}}
```
Let's add a custom bucket histogram with negative observations and negative bucket:
```
cbh_has_neg {{schema:-53 sum:172.5 count:15 custom_values:[-10 5 10 15 20] buckets:[2 1 6 4 3 1]}}
```

krajorama commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -1373,3 +1373,60 @@ eval instant at 1m histogram_fraction(-Inf, +Inf, histogram_nan)
+
+eval instant at 1m cbh >/ 15
+{__name__="cbh"} {{schema:-53 count:4 sum:72.5 custom_values:[5 10 15 20] offset:3 buckets:[3 1]}}
+
```
Let's also add some tests for negative buckets with cbh:
```
eval instant at 1m cbh_has_neg </ 2
  cbh_has_neg{} {{schema:-53 count:2.8 sum:5.5 custom_values:[-10 5 10 15 20] buckets:[2 0.8]}}

eval instant at 1m cbh_has_neg </ -4
  cbh_has_neg{} {{schema:-53 count:2.4 sum:6.5 custom_values:[-10 5 10 15 20] buckets:[2 0.4]}}

eval instant at 1m cbh_has_neg </ -15
  cbh_has_neg{} {{schema:-53 count:1 sum:17.5 custom_values:[-10 5 10 15 20] buckets:[1]}}
```

I've made these with a fix to the comment about the interpolation for -Inf

sujalshah-bit commented:
@krajorama can you kindly review :)

krajorama commented on promql/engine.go:
```diff
@@ -2900,6 +2900,309 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+func computeExponentialTrim(bucket histogram.Bucket[float64], rhs float64, isPositive bool, op parser.ItemType) (float64, float64) {
+	var fraction, bucketMidpoint, keepCount float64
+
+	logLower := math.Log2(math.Abs(bucket.Lower))
```
Technically exponential buckets can also have -Inf and +Inf lower and upper bounds. See 
https://prometheus.io/docs/specs/native_histograms/#buckets
and
https://github.com/prometheus/prometheus/blob/ffcba01c5a4e7f5f35a15a30ba283409e8cd0138/model/histogram/generic.go#L470

krajorama commented on promql/engine.go:
```diff
@@ -2900,6 +2900,309 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+}
+
+// Helper function to trim native histogram buckets.
+func trimHistogram(trimmedHist *histogram.FloatHistogram, rhs float64, op parser.ItemType) {
```
This function and the compute*Trim functions got a little bit complicated. Also as I note later, the computeExponentialTrim function is missing handling of overflow -Inf, +Inf buckets.

I'd like to suggest a simplification (we can discuss putting it in a follow-up PR)
let's have a function that has this signature:
```
computeSplit(bucket histogram.Bucket[float64], le float64, isPostive, isCustomBucket bool) (underCount, bucketMidPoint)
```

This would return the portion of the bucket count that's under or equal to the le and the midPoint of the bucket values (if possible).

Note that this doesn't have to take the Op. Since the result is directly usable as keepCount for TRIM_UPPER and you just need to subtract the underCount from the bucket count to get the keepCount value for TRIM_LOWER.

Also this would make the algorithm closer to https://github.com/prometheus/prometheus/blob/ffcba01c5a4e7f5f35a15a30ba283409e8cd0138/promql/quantile.go#L386 which does something very similar.

Also also, the special handling for (-Inf, positive] bucket for custom buckets will never kick in for exponential histograms, since they don't have such bucket.

The resulting function should be possible to reuse here easily.

krajorama reviewed (commented):
I think this is close to being merge-able. And maybe we should merge it to get a good baseline. I made one comment about TRIM_LOWER vs +Inf bucket which I think we should fix before merging. And more comments about exponential buckets and general structure which we could push later. wdyt @beorn7 ?

beorn7 commented:
Thanks, everyone. I'll give this a thorough review ASAP (hopefully tomorrow).

sujalshah-bit commented:
Should I go ahead and address [this comment](https://github.com/prometheus/prometheus/pull/16330#discussion_r2190083052) and [this one](https://github.com/prometheus/prometheus/pull/16330#discussion_r2190156376) in this PR, or should I work on those as a follow-up instead?

Either way, I’m happy to take care of it. Please let me know what works best.

Aside from those, I noticed only one other minor [fix](https://github.com/prometheus/prometheus/pull/16330#discussion_r2189918285) remaining (just have to push).

beorn7 commented:
I was back in action for a few days, but now I'm going on vacations… @krajorama @sujalshah-bit if you find the time, please finish this without me. I'm taking a note that I'll check the logic of this in any case before we declare NH stable.

sujalshah-bit commented:
@krajorama I have made the changes you can have look.

krajorama commented on promql/engine.go:
```diff
@@ -2903,6 +2903,263 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+	product := math.Abs(b.Lower) * math.Abs(rhs)
+	underCount := computeSplit(b, rhs, isPositive, isCustomBucket)
+	if op == parser.TRIM_UPPER {
+		return underCount, computeMidpoint(b, product, isCustomBucket, isPositive)
```
It seems very suspicious that for custom buckets we use the Lower and Upper bound for the mid point, but for exponential buckets we go from the lower bound to the rhs (or rhs to upper bound) for the mid point calculation.

Feels like one of those is wrong.

krajorama commented on promql/engine.go:
```diff
@@ -2903,6 +2903,263 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
+	trimmedHist.Compact(0)
+}
+
+func computeMidpoint(b histogram.Bucket[float64], product float64, isCustom, isPositive bool) float64 {
```
Following from my comment above I would expect this function to simply look like this:
```
func computeMidpoint(a, b float64, isLinear, isPositive bool) float64 {
	if isLinear {
		return (b + a) / 2
	}

	if isPositive {
		return math.Sqrt(math.Abs(a) * math.Abs(b))
	}
	return -math.Sqrt(math.Abs(a) * math.Abs(b))
}
```

But I could not get every test to pass. So I think that question needs answering about why we use different kind of ranges when calling this function. Bucket lower/upper bound for custom buckets, but range adjusted with `rhs` for the exponential buckets.

beorn7 commented:
@sujalshah-bit just to let you know: @krajorama is on vacations for the rest of the month, but I'll continue the review once you have addressed @krajorama's latest comments. Just leave a note here once the PR is ready for the next round of review.

linasm commented:
> Yes, I am planning to do it, but not any time soon. If someone wants to pick the issue, please feel free to take the PR forward.

@sujalshah-bit (also @beorn7 and @krajorama ) I see no progress on this PR for almost 6 months now. If there are no objections, I can pick it up and try taking it across the finish line.
I would start by addressing the existing comments.
Also, after looking a the code, I am tempted to make `trimHistogram` a method of `FloatHistogram`, where it would fit nicely  with other histogram operations (`Mul`, `Div`, etc.). Moving it (and its helper functions) to https://github.com/prometheus/prometheus/blob/main/model/histogram/float_histogram.go would avoid polluting `engine.go` (which already is so huge) with few hundred LoC of histogram specific code. This move would be done separately from addressing the existing comments.
